src/tensor_ops/cmp/mod.rs

```diff
@@ -0,0 +1,458 @@
+use crate::{
+    gradients::{NoneTape, Tape},
+    shapes::{Shape, Unit},
+    tensor::{DeviceStorage, Tensor},
+};
+
+mod cpu_kernels;
+#[cfg(feature = "cuda")]
+mod cuda_kernels;
+
+pub trait CmpKernel<Op, E: Unit>: DeviceStorage {
+    fn forward<S: Shape>(
+        &self,
+        lhs: &Self::Storage<S, E>,
+        rhs: &Self::Storage<S, E>,
+    ) -> Result<Self::Storage<S, bool>, Self::Err>;
+}
+
+fn try_cmp_op<Op, S: Shape, E: Unit, D: CmpKernel<Op, E>, T: Tape<D>>(
+    lhs: &Tensor<S, E, D, T>,
+    rhs: &Tensor<S, E, D, T>,
+) -> Result<Tensor<S, bool, D, NoneTape>, D::Err> {
```

Comment:
Will need to add a runtime shape check as I discovered in #420 
```suggestion
) -> Result<Tensor<S, bool, D, NoneTape>, D::Err> {
    assert_eq!(lhs.shape(), rhs.shape());
```